### ISSUE TITLE
[QUESTION] How can I get default value In Custom Component

### ISSUE THREAD

**Issue by pummy-kanhasoft** [formio/formio.js#4725]
## Environment

- Hosting type
  - [ ] Form.io
  - [ ] Local deployment
    - Version: 
- Formio.js version: 4.14.5
- Frontend framework: Next Js, React Js
- Browser: Chrome, Mozilla
- Browser version:

## Expected behavior
I need to get default value in custom component.
The below screenshot is custom component of currency convertor. When I select currency from and currency to in default value, I don't get data in right side preview section. How can I get this default value ?
https://drive.google.com/file/d/15RqIHCBsUWQSh0uTGhRQF-eOUTYznDLa/view?usp=sharing
https://drive.google.com/file/d/1sYcUj39WyG2jFwCNBDwlvwZoIk2QqakB/view?usp=sharing

## Expected output
https://drive.google.com/file/d/1pom84h5ScCZhlwOs9-mrPP9hXx-KfNHY/view?usp=sharing

## Observed output
https://nimb.ws/N5yuh7

```js
//code here
import { Components } from "react-formio";
const Input = Components.components.input;
const FieldComponent = Components.components.field;
const editForm = Components.components.nested.editForm;
import $ from "jquery";

export default class MyComponent extends FieldComponent {
  /**
   * This is the default schema of your custom component. It will "derive"
   * from the base class "schema" and extend it with its default JSON schema
   * properties. The most important are "type" which will be your component
   * type when defining new components.
   *
   * @param extend - This allows classes deriving from this component to
   *                 override the schema of the overridden class.
   */
  static schema(...extend) {
    return FieldComponent.schema(
      {
        type: "customcurrency",
        label: "Custom Currency Convertor",
        key: "customcurrency",
        amount: "amount",
        amount_from: "amount_from",
        amount_to: "amount_to",
      },
      ...extend
    );
  }

  /**
   * This is the Form Builder information on how this component should show
   * up within the form builder. The "title" is the label that will be given
   * to the button to drag-and-drop on the buidler. The "icon" is the font awesome
   * icon that will show next to it, the "group" is the component group where
   * this component will show up, and the weight is the position within that
   * group where it will be shown. The "schema" field is used as the default
   * JSON schema of the component when it is dragged onto the form.
   */
  static get builderInfo() {
    return {
      title: "Custom Currency Convertor",
      icon: "dollar",
      group: "basic",
      documentation: "/userguide/#textfield",
      weight: 0,
      schema: MyComponent.schema(),
    };
  }

  /**
   * Called when the component has been instantiated. This is useful to define
   * default instance variable values.
   *
   * @param component - The JSON representation of the component created.
   * @param options - The global options for the renderer
   * @param data - The contextual data object (model) used for this component.
   */
  constructor(component, options, data) {
    super(component, options, data);
  }

  /**
   * Called immediately after the component has been instantiated to initialize
   * the component.
   */
  init() {
    super.init();
  }

  /**
   * For Input based components, this returns the <input> attributes that should
   * be added to the input elements of the component. This is useful if you wish
   * to alter the "name" and "class" attributes on the <input> elements created
   * within this component.
   *
   * @return - A JSON object that is the attribute information to be added to the
   *           input element of a component.
   */
  get inputInfo() {
    return super.inputInfo;
  }

  /**
   * This method is used to render a component as an HTML string. This method uses
   * the template system (see Form Templates documentation) to take a template
   * and then render this as an HTML string.
   *
   * @param content - Important for nested components that receive the "contents"
   *                  of their children as an HTML string that should be injected
   *                  in the {{ content }} token of the template.
   *
   * @return - An HTML string of this component.
   */
  render(_content) {
    let table = "<div>";
    table +=
      '<input type ="text" ref="currencytext_amount" value="" id="currencyconverter_amount" placeholder="Enter Amount" class="form-control cls_currencyconverter_amount">';
    table += "</div><div></br>";
    table +=
      "<select ref='currencytext_amount_from' class='form-select currency_from_amount' aria-label='From' id='from' ><option >Select Currency From</option><option value='USD'>USD</option><option value='AED'>AED</option><option value='ARS'>ARS</option><option value='AUD'>AUD</option><option value='BGN'>BGN</option><option value='BRL'>BRL</option><option value='BSD'>BSD</option><option value='CAD'>CAD</option><option value='CHF'>CHF</option><option value='CLP'>CLP</option><option value='CNY'>CNY</option><option value='COP'>COP</option><option value='CZK'>CZK</option><option value='DKK'>DKK</option><option value='DOP'>DOP</option><option value='EGP'>EGP</option><option value='EUR'>EUR</option><option value='FJD'>FJD</option><option value='GBP'>GBP</option><option value='GTQ'>GTQ</option><option value='HKD'>HKD</option><option value='HRK'>HRK</option><option value='HUF'>HUF</option><option value='IDR'>IDR</option><option value='ILS'>ILS</option><option value='INR'>INR</option><option value='ISK'>ISK</option><option value='JPY'>JPY</option><option value='KRW'>KRW</option><option value='KZT'>KZT</option><option value='MVR'>MVR</option><option value='MXN'>MXN</option><option value='MYR'>MYR</option><option value='NOK'>NOK</option><option value='NZD'>NZD</option><option value='PAB'>PAB</option><option value='PEN'>PEN</option><option value='PHP'>PHP</option><option value='PKR'>PKR</option><option value='PLN'>PLN</option><option value='PYG'>PYG</option><option value='RON'>RON</option><option value='RUB'>RUB</option><option value='SAR'>SAR</option><option value='SEK'>SEK</option><option value='SGD'>SGD</option><option value='THB'>THB</option><option value='TRY'>TRY</option><option value='TWD'>TWD</option><option value='UAH'>UAH</option><option value='UYU'>UYU</option><option value='ZAR'>ZAR</option></select>";
    table += "</div><div><br/>";
    table +=
      "<select ref='currencytext_amount_to' class='form-select currency_to_amount' aria-label='From' id='from' ><option >Select Currency To</option><option value='USD'>USD</option><option value='AED'>AED</option><option value='ARS'>ARS</option><option value='AUD'>AUD</option><option value='BGN'>BGN</option><option value='BRL'>BRL</option><option value='BSD'>BSD</option><option value='CAD'>CAD</option><option value='CHF'>CHF</option><option value='CLP'>CLP</option><option value='CNY'>CNY</option><option value='COP'>COP</option><option value='CZK'>CZK</option><option value='DKK'>DKK</option><option value='DOP'>DOP</option><option value='EGP'>EGP</option><option value='EUR'>EUR</option><option value='FJD'>FJD</option><option value='GBP'>GBP</option><option value='GTQ'>GTQ</option><option value='HKD'>HKD</option><option value='HRK'>HRK</option><option value='HUF'>HUF</option><option value='IDR'>IDR</option><option value='ILS'>ILS</option><option value='INR'>INR</option><option value='ISK'>ISK</option><option value='JPY'>JPY</option><option value='KRW'>KRW</option><option value='KZT'>KZT</option><option value='MVR'>MVR</option><option value='MXN'>MXN</option><option value='MYR'>MYR</option><option value='NOK'>NOK</option><option value='NZD'>NZD</option><option value='PAB'>PAB</option><option value='PEN'>PEN</option><option value='PHP'>PHP</option><option value='PKR'>PKR</option><option value='PLN'>PLN</option><option value='PYG'>PYG</option><option value='RON'>RON</option><option value='RUB'>RUB</option><option value='SAR'>SAR</option><option value='SEK'>SEK</option><option value='SGD'>SGD</option><option value='THB'>THB</option><option value='TRY'>TRY</option><option value='TWD'>TWD</option><option value='UAH'>UAH</option><option value='UYU'>UYU</option><option value='ZAR'>ZAR</option></select>";
    table += "</div><div><br/>";
    
    table +=
      '<input type ="number"  id="currencyconverter_result" placeholder="Converted Value" class="form-control cls_currencyconverter_result" readonly>';
    table += "</div>";
    return super.render(table);
  }

  /**
   * The attach method is called after "render" which takes the rendered contents
   * from the render method (which are by this point already added to the DOM), and
   * then "attach" this component logic to that html. This is where you would load
   * any references within your templates (which use the "ref" attribute) to assign
   * them to the "this.refs" component variable (see comment below).
   *
   * @param - The parent DOM HtmlElement that contains the component template.
   *
   * @return - A Promise that will resolve when the component has completed the
   *           attach phase.
   */
  attach(element) {
    /**
     * This method will look for an element that has the 'ref="customRef"' as an
     * attribute (like <div ref="customRef"></div>) and then assign that DOM
     * element to the variable "this.refs". After this method is executed, the
     * following will point to the DOM element of that reference.
     *
     * this.refs.customRef
     *
     * For DOM elements that have multiple in the component, you would make this
     * say 'customRef: "multiple"' which would then turn "this.refs.customRef" into
     * an array of DOM elements.
     */
    this.loadRefs(element, {
      currencytext_amount: "single",
      currencytext_amount_from: "single",
      currencytext_amount_to: "single",
    });

    /**
     * It is common to attach events to your "references" within your template.
     * This can be done with the "addEventListener" method and send the template
     * reference to that object.
     */

    // amount value
    this.addEventListener(this.refs.currencytext_amount, "input", () => {
      this.amountValue = document.getElementsByClassName("cls_currencyconverter_amount")[0].value;
      this.updateValue();
    });


    //amount from
    this.addEventListener(this.refs.currencytext_amount_from, "change", () => {
      this.amount_from = document.getElementsByClassName(
        "currency_from_amount"
      )[0].value;
      this.updateValue();
    });

    // amount to
    this.addEventListener(this.refs.currencytext_amount_to, "change", () => {
      this.amount_to =
        document.getElementsByClassName("currency_to_amount")[0].value;
      this.updateValue();
    });

    return super.attach(element);
  }

  /**
   * Called when the component has been detached. This is where you would destroy
   * any other instance variables to free up memory. Any event registered with
   * "addEventListener" will automatically be detached so no need to remove them
   * here.
   *
   * @return - A Promise that resolves when this component is done detaching.
   */
  detach() {
   
    return super.detach();
  }

  /**
   * Called when the component has been completely "destroyed" or removed form the
   * renderer.
   *
   * @return - A Promise that resolves when this component is done being destroyed.
   */
  destroy() {
   
    return super.destroy();
  }

  /**
   * A very useful method that will take the values being passed into this component
   * and convert them into the "standard" or normalized value. For exmample, this
   * could be used to convert a string into a boolean, or even a Date type.
   *
   * @param value - The value that is being passed into the "setValueAt" method to normalize.
   * @param flags - Change propogation flags that are being used to control behavior of the
   *                change proogation logic.
   *
   * @return - The "normalized" value of this component.
   */
  normalizeValue(_value, flags = {}) {
    let newValue = this.final_value;
    return super.normalizeValue(newValue, flags);
  }

  /**
   * Returns the value of the "view" data for this component.
   *
   * @return - The value for this whole component.
   */
  getValue() {

    const api = "https://api.exchangerate-api.com/v4/latest/USD";
    fetch(`${api}`)
      .then((currency) => {
        return currency.json();
      })
      .then((response) => {

        let data = response;
        let fromRate1 = data.rates[this.amount_from];
        let toRate1 = data.rates[this.amount_to];
        let value = ((toRate1 / fromRate1) * this.amountValue).toFixed(2);

        this.final_value = value;
        document.getElementById("currencyconverter_result").value = this.final_value;
      });

    return super.getValue();
  }

  /**
   * Much like "getValue", but this handles retrieving the value of a single index
   * when the "multiple" flag is used within the component (which allows them to add
   * multiple values). This turns a single value into an array of values, and this
   * method provides access to a certain index value.
   *
   * @param index - The index within the array of values (from the multiple flag)
   *                that is getting fetched.
   *
   * @return - The view data of this index.
   */
  getValueAt(index) {
    return super.getValueAt(index);
  }

  /**
   * Sets the value of both the data and view of the component (such as setting the
   * <input> value to the correct value of the data. This is most commonly used
   * externally to set the value and also see that value show up in the view of the
   * component. If you wish to only set the data of the component, like when you are
   * responding to an HMTL input event, then updateValue should be used instead since
   * it only sets the data value of the component and not the view.
   *
   * @param value - The value that is being set for this component's data and view.
   * @param flags - Change propogation flags that are being used to control behavior of the
   *                change proogation logic.
   *
   * @return - Boolean indicating if the setValue changed the value or not.
   */
  setValue(_value, flags = {}) {
    let setValue = this.final_value;
    document.getElementsByClassName("currency_to_amount")[0].value;
    return super.setValue(setValue, flags);
  }

  /**
   * Sets the value for only this index of the component. This is useful when you have
   * the "multiple" flag set for this component and only wish to tell this component
   * how the value should be set on a per-row basis.
   *
   * @param index - The index within the value array that is being set.
   * @param value - The value at this index that is being set.
   * @param flags - Change propogation flags that are being used to control behavior of the
   *                change proogation logic.
   *
   * @return - Boolean indiciating if the setValue at this index was changed.
   */
  setValueAt(index, _value, flags = {}) {
    let setValAt = this.final_value;
    return super.setValueAt(index, setValAt, flags);
  }

  /**
   * Similar to setValue, except this does NOT update the "view" but only updates
   * the data model of the component.
   *
   * @param value - The value of the component being set.
   * @param flags - Change propogation flags that are being used to control behavior of the
   *                change proogation logic.
   *
   * @return - Boolean indicating if the updateValue changed the value or not.
   */

  updateValue(_value, _flags = {}) {
    return super.updateValue();
  }
}

```

**Comment by paulbors** [comment]
Any reason why you would expect this to return your custom component's value?
return super.getValue();

Or this?
return super.getValueAt(index);

And the same goes for the setter:
return super.setValue(setValue, flags);

or this:
return super.setValueAt(index, setValAt, flags);

Why don't you read the online doc on how to write a custom component?
https://docs.form.io/developers/custom-components

And don't say the example calls super, and so did I because... You should know better :)

**Comment by pummy-kanhasoft** [comment]
Hi @paulbors 

Below link whick is demo of what I need.
![image](https://user-images.githubusercontent.com/91519891/174739352-efcb226f-5741-48bb-8c0d-9bba00e80cf7.png)


This below link is screenshot which I need is Data->Default Value on right side. 
https://www.awesomescreenshot.com/image/29027937?key=b6c7195f6a61b3ddf157d0fcae8bf6b2

I need left side of box "Default value" dropdown value on right side of dropdown.

By using this below link of cutom component I created my custom component.
https://docs.form.io/developers/custom-components

And also I tried getting default value from popup of checkmatrix custom component, but its now working.

**Comment by pummy-kanhasoft** [comment]
Hi @paulbors 

Whet I need is given in the below screenshot.
https://tinyurl.com/248hqy24

Below link screenshot which I need Data->Default value on right side.
https://www.awesomescreenshot.com/image/29030353?key=b82bccfe8d7ce676f4432cdefdfc4f5d

I need left side of box "Default Value" dropdown value on right side of dropdown.
By using this below link of custom component, I created my custom component.
https://docs.form.io/developers/custom-components

And also I tried getting default value from popup of check matrix custom component, but it's not working,
